vendor/bat-native-ledger/src/bat/ledger/internal/promotion.cc

```diff
@@ -0,0 +1,851 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ledger/internal/promotion.h"
+
+#include <map>
+#include <memory>
+#include <utility>
+
+#include "base/json/json_reader.h"
+#include "base/json/json_writer.h"
+#include "base/strings/stringprintf.h"
+#include "base/time/time.h"
+#include "bat/ledger/internal/bat_util.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/bat_helper.h"
+#include "bat/ledger/internal/state_keys.h"
+#include "bat/ledger/internal/static_values.h"
+#include "bat/ledger/internal/request/promotion_requests.h"
+#include "bat/ledger/internal/request/request_util.h"
+#include "bat/ledger/internal/common/bind_util.h"
+#include "bat/ledger/internal/common/security_helper.h"
+#include "bat/ledger/internal/common/time_util.h"
+#include "brave_base/random.h"
+#include "net/http/http_status_code.h"
+
+#include "wrapper.hpp"  // NOLINT
+
+using std::placeholders::_1;
+using std::placeholders::_2;
+using std::placeholders::_3;
+
+using challenge_bypass_ristretto::BatchDLEQProof;
+using challenge_bypass_ristretto::BlindedToken;
+using challenge_bypass_ristretto::PublicKey;
+using challenge_bypass_ristretto::SignedToken;
+using challenge_bypass_ristretto::UnblindedToken;
+
+namespace braveledger_promotion {
+
+Promotion::Promotion(bat_ledger::LedgerImpl* ledger) :
+    attestation_(std::make_unique<braveledger_attestation::AttestationImpl>
+        (ledger)),
+    ledger_(ledger) {
+}
+
+Promotion::~Promotion() {
+}
+
+std::string ParseOSToString(ledger::OperatingSystem os) {
+  switch (static_cast<int>(os)) {
+    case static_cast<int>(ledger::OperatingSystem::WINDOWS):  {
+      return "windows";
+    }
+    case static_cast<int>(ledger::OperatingSystem::MACOS):  {
+      return "osx";
+    }
+    case static_cast<int>(ledger::OperatingSystem::LINUX):  {
+      return "linux";
+    }
+    case static_cast<int>(ledger::OperatingSystem::UNDEFINED):  {
+      return "undefined";
+    }
+    default: {
+      NOTREACHED();
+      return "";
+    }
+  }
+}
+
+std::string ParseClientInfoToString(ledger::ClientInfoPtr info) {
+  if (!info) {
+    return "";
+  }
+
+  switch (static_cast<int>(info->platform)) {
+    case static_cast<int>(ledger::Platform::ANDROID):  {
+      return "android";
+    }
+    case static_cast<int>(ledger::Platform::IOS):  {
+      return "ios";
+    }
+    case static_cast<int>(ledger::Platform::DESKTOP):  {
+      return ParseOSToString(info->os);
+    }
+    default: {
+      NOTREACHED();
+      return "";
+    }
+  }
+}
+
+ledger::PromotionType ConvertStringToPromotionType(const std::string& type) {
+  if (type == "ugp") {
+    return ledger::PromotionType::UGP;
+  }
+
+  if (type == "ads") {
+    return ledger::PromotionType::ADS;
+  }
+
+  // unknown promotion type, returning dummy value.
+  NOTREACHED();
+  return ledger::PromotionType::UGP;
```

Comment:
it needs to be a valid value as it would not compile. But `NOTREACHED` will make sure that we know that there is a problem so this value will never be used